sql: ensure auto-retrying transactions respect the statement_timeout

arulajmani opened the pull request:
Backport:
  * 1/1 commits from "sql: ensure auto-retrying transactions respect the statement_timeout" (#53968)
  * 1/1 commits from "sql: increase statement_timeout in sequences logictest" (#54368)

Please see individual PRs for details.

/cc @cockroachdb/release

arulajmani committed:
sql: ensure auto-retrying transactions respect the statement_timeout

Previously, if a statement exceeded the statement_timnout in the
conn_executor, but failed with a retryable error, we would still retry
it. This was because even though we overrode the error, we didn't
override the `fms.Event` and `fsm.EventPayload` which actually dictates
the next transition. This patch overrides them, thereby ensuring that
even if a timed out query/cancelled query encountered a retryable
error, we do not transition into retrying it.

Fixes #52845

Release justification: low risk, high benefit changes to existing functionality

Release note (bug fix): queries that can be automatically retried did
not respect the `statement_timeout` earlier, which is now fixed.

arulajmani committed:
sql: increase statement_timeout in sequences logictest

Previously the statement timeout in a logic test was set to 1ms, which
was sometimes not enough to complete resetting the statement timeout
after the test, causing it to flake. This PR increases the timeout
from 1ms to 10ms so that we don't flake on the cleanup.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54370)
<!-- Reviewable:end -->

rafiss commented on pkg/sql/run_control_test.go:
```diff
@@ -850,6 +849,42 @@ func TestIdleInTransactionSessionTimeoutCommitWaitState(t *testing.T) {
+	defer tc.Stopper().Stop(ctx)
+
+	conn, err := tc.ServerConn(0).Conn(ctx)
+	require.NoError(t, err)
```
looks like the `require` import was added in 39e7448ccddf92428716fbe56d257edaee02c874

i think you can just add the import here

arulajmani committed:
sql: exempt `SET` statements from the statement_timeout

Previously, if a user set a really low statement_timeout value, there
would be no way to reset it/remove the statement_timeout entirely. To
get around this, all `SET` statements are now exempt from the statement
timeout.This should also fix some of the flakes we were seeing in
checks were decoupled in #53968. `SET` statements aren't canceled as no
one checks for a canceled context, which meant this exemption existed
implicitly before #53968.

Closes #54372

Release note: None

arulajmani commented:
I might need to get https://github.com/cockroachdb/cockroach/issues/54372 in as well to deflake CI, so I'm putting the DNM label on this for now.

arulajmani commented:
This should be ready now that #54415  is in.
